File: api/app/schemas/core_study.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.schemas.integration import TopicScoreResponse


CoreEventName = Literal[
    "onboarding_started",
    "onboarding_completed",
    "course_created",
    "material_uploaded",
    "quiz_started",
    "quiz_submitted",
    "flashcard_reviewed",
    "course_workspace_viewed",
    "course_workspace_tab_changed",
    "study_plan_viewed",
    "study_plan_refreshed",
    "study_plan_item_started",
    "study_plan_item_completed",
    "study_plan_item_dismissed",
    "remediation_created",
]


class ProductEventCreate(BaseModel):
    event_id: str = Field(min_length=1, max_length=128)
    name: CoreEventName
    properties: dict[str, Any] = Field(default_factory=dict)
    occurred_at: datetime


class ProductEventBatchRequest(BaseModel):
    events: list[ProductEventCreate] = Field(min_length=1, max_length=100)


class ProductEventBatchResponse(BaseModel):
    accepted: int
    duplicates: int


class StudyProfilePatch(BaseModel):
    daily_minutes: int | None = Field(default=None, ge=5, le=480)
    study_goal: str | None = Field(default=None, max_length=255)
    onboarding_completed: bool | None = None


class OnboardingCourseCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    exam_date: datetime
    confidence_level: Literal["low", "medium", "high"]


class OnboardingChecklist(BaseModel):
    profile: bool
    course: bool
    material: bool
    quiz: bool
    attempt: bool
    complete: bool


class StudyProfileResponse(BaseModel):
    daily_minutes: int
    study_goal: str | None = None
    onboarding_completed_at: datetime | None = None


class OnboardingResponse(BaseModel):
    profile: StudyProfileResponse
    checklist: OnboardingChecklist


class WorkspaceResource(BaseModel):
    id: UUID
    title: str
    status: str | None = None
    count: int | None = None
    updated_at: datetime | None = None


class CourseWorkspaceResponse(BaseModel):
    course_id: UUID
    title: str
    confidence_level: str | None = None
    exam_date: datetime | None = None
    materials: list[WorkspaceResource]
    quizzes: list[WorkspaceResource]
    flashcard_decks: list[WorkspaceResource]
    progress: dict[str, Any]


class StudyPlanItemResponse(BaseModel):
    id: UUID
    course_id: UUID | None = None
    plan_date: str
    item_type: str
    title: str
    topic: str | None = None
    target_id: UUID | None = None
    estimated_minutes: int
    priority: int
    status: str
    started_at: datetime | None = None
    completed_at: datetime | None = None
    dismissed_at: datetime | None = None


class TodayStudyPlanResponse(BaseModel):
    date: str
    daily_minutes: int
    generated: bool
    items: list[StudyPlanItemResponse]


class StudyPlanItemPatch(BaseModel):
    action: Literal["start", "complete", "dismiss"]


class StudyPlanItemCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    estimated_minutes: int = Field(default=15, ge=5, le=180)
    item_type: Literal["custom", "flashcards", "weak_topic", "course_review"] = "custom"
    course_id: UUID | None = None
    topic: str | None = Field(default=None, max_length=255)
    target_id: UUID | None = None


class RemediationRequest(BaseModel):
    action: Literal["deck", "retry", "explain"]
    topics: list[str] = Field(default_factory=list, max_length=20)

    @model_validator(mode="after")
    def normalize_topics(self) -> "RemediationRequest":
        self.topics = sorted({topic.strip() for topic in self.topics if topic.strip()})
        return self


class RemediationResponse(BaseModel):
    attempt_id: UUID
    action: Literal["deck", "retry", "explain"]
    topics: list[str]
    status: str
    resource_id: UUID | None = None
    url: str
    context: dict[str, Any] = Field(default_factory=dict)


class RemediationOptionsResponse(BaseModel):
    attempt_id: UUID
    weak_topics: list[str]
    available_actions: list[Literal["deck", "retry", "explain"]]
    existing: list[RemediationResponse] = Field(default_factory=list)


class FeedResourceMeta(BaseModel):
    question_count: int | None = None
    card_count: int | None = None
    status: str | None = None
    score_pct: float | None = None
    file_name: str | None = None
    chunk_count: int | None = None


class FeedResourceItemResponse(BaseModel):
    kind: Literal["quiz", "deck", "material"]
    id: UUID
    title: str
    course_id: UUID | None = None
    course_title: str | None = None
    last_activity_at: datetime
    activity_label: str
    meta: FeedResourceMeta = Field(default_factory=FeedResourceMeta)
    preview: list[str] = Field(default_factory=list)
    href: str


class CourseSummaryResponse(BaseModel):
    id: UUID
    title: str
    exam_date: datetime | None = None


class StudyFeedResponse(BaseModel):
    next_exam: CourseSummaryResponse | None = None
    exam_items: list[FeedResourceItemResponse] = Field(default_factory=list)
    recent_items: list[FeedResourceItemResponse] = Field(default_factory=list)
    weak_topics: list[TopicScoreResponse] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)

File: client/lib/core-study.ts
import { apiRequest } from "@/lib/api";
import type { Course } from "@/lib/study";

export type ConfidenceLevel = "low" | "medium" | "high";

export type OnboardingChecklist = {
  profile: boolean;
  course: boolean;
  material: boolean;
  quiz: boolean;
  attempt: boolean;
  complete: boolean;
};

export type OnboardingState = {
  profile: {
    daily_minutes: number;
    study_goal: string | null;
    onboarding_completed_at: string | null;
  };
  checklist: OnboardingChecklist;
};

export type OnboardingProfilePatch = {
  daily_minutes?: number;
  study_goal?: string;
  onboarding_completed?: boolean;
};

export type OnboardingCourseInput = {
  title: string;
  description?: string;
  exam_date?: string;
  confidence_level?: ConfidenceLevel;
};

export type WorkspaceResource = {
  id: string;
  title: string;
  status: string | null;
  count: number | null;
  updated_at: string | null;
};

export type CourseWorkspace = {
  course_id: string;
  title: string;
  confidence_level: string | null;
  exam_date: string | null;
  materials: WorkspaceResource[];
  quizzes: WorkspaceResource[];
  flashcard_decks: WorkspaceResource[];
  progress: Record<string, unknown>;
};

export type StudyPlanAction = "start" | "complete" | "dismiss";

export type StudyPlanItemCreate = {
  title: string;
  estimated_minutes?: number;
  item_type?: "custom" | "flashcards" | "weak_topic" | "course_review";
  course_id?: string | null;
  topic?: string | null;
  target_id?: string | null;
};

export type StudyPlanItem = {
  id: string;
  course_id: string | null;
  plan_date: string;
  item_type: string;
  title: string;
  topic: string | null;
  target_id: string | null;
  estimated_minutes: number;
  priority: number;
  status: string;
  started_at: string | null;
  completed_at: string | null;
  dismissed_at: string | null;
};

export type TodayStudyPlan = {
  date: string;
  daily_minutes: number;
  generated: boolean;
  items: StudyPlanItem[];
};

export type FeedResourceMeta = {
  question_count?: number | null;
  card_count?: number | null;
  status?: string | null;
  score_pct?: number | null;
  file_name?: string | null;
  chunk_count?: number | null;
};

export type FeedResourceItem = {
  kind: "quiz" | "deck" | "material";
  id: string;
  title: string;
  course_id: string | null;
  course_title: string | null;
  last_activity_at: string;
  activity_label: string;
  meta: FeedResourceMeta;
  preview: string[];
  href: string;
};

export type CourseSummary = {
  id: string;
  title: string;
  exam_date: string | null;
};

export type StudyFeed = {
  next_exam: CourseSummary | null;
  exam_items: FeedResourceItem[];
  recent_items: FeedResourceItem[];
  weak_topics: { topic: string; score_pct: number }[];
  recommendations: string[];
};

export type RemediationAction = "deck" | "retry" | "explain";

export type RemediationResult = {
  attempt_id: string;
  action: RemediationAction;
  topics: string[];
  status: string;
  resource_id: string | null;
  url: string;
  context: Record<string, unknown>;
};

export type RemediationOptions = {
  attempt_id: string;
  weak_topics: string[];
  available_actions: RemediationAction[];
  existing: RemediationResult[];
};

export function getOnboarding(token: string) {
  return apiRequest<OnboardingState>("/onboarding", { token });
}

export function updateOnboarding(token: string, patch: OnboardingProfilePatch) {
  return apiRequest<OnboardingState>("/onboarding", {
    method: "PATCH",
    token,
    body: JSON.stringify(patch)
  });
}

export function createOnboardingCourse(token: string, input: OnboardingCourseInput) {
  return apiRequest<Course>("/onboarding/course", {
    method: "POST",
    token,
    body: JSON.stringify(input)
  });
}

export function getCourseWorkspace(token: string, courseId: string) {
  return apiRequest<CourseWorkspace>(`/courses/${courseId}/workspace`, { token });
}

export function getTodayStudyPlan(token: string) {
  return apiRequest<TodayStudyPlan>("/study/today", { token });
}

export function getStudyFeed(token: string) {
  return apiRequest<StudyFeed>("/study/feed", { token });
}

export function refreshTodayStudyPlan(token: string) {
  return apiRequest<TodayStudyPlan>("/study/today/refresh", { method: "POST", token });
}

export function updateStudyPlanItem(token: string, itemId: string, action: StudyPlanAction) {
  return apiRequest<StudyPlanItem>(`/study/plan-items/${itemId}`, {
    method: "PATCH",
    token,
    body: JSON.stringify({ action })
  });
}

export function createStudyPlanItem(token: string, input: StudyPlanItemCreate) {
  return apiRequest<StudyPlanItem>("/study/plan-items", {
    method: "POST",
    token,
    body: JSON.stringify(input)
  });
}

export function getRemediationOptions(token: string, attemptId: string) {
  return apiRequest<RemediationOptions>(`/quizzes/attempts/${attemptId}/remediation`, { token });
}

export function createRemediation(
  token: string,
  attemptId: string,
  action: RemediationAction,
  topics: string[] = []
) {
  return apiRequest<RemediationResult>(`/quizzes/attempts/${attemptId}/remediation`, {
    method: "POST",
    token,
    body: JSON.stringify({ action, topics })
  });
}
